### PR TITLE
Throw errors for invalid doc

### PR DESCRIPTION
**Pull request by fisker**
## Description

The printer is now async, it can easily forget to `await` it, should refuse to print if `Promise` inside doc.

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Review by kachkaev — APPROVED** [review]

**Comment by thorn0** [comment]
How about making a `getDocType` function, which could be reused in `switch` statements in `fits` and `printDocToString`? The function would return `doc.type` or special values: `string`, `array`, `invalid`. `invalid` is returned if the argument isn't a string, an array, or an object with `type`.

**Comment by fisker** [comment]
I was going to do that when we remove `concat`

**Review by sosukesuzuki — APPROVED** [review]
I don't disagree with thorn0, but personally I am happy with this PR!

File: src/document/printer.js [hunk 1]
```diff
@@ -541,7 +541,7 @@ function printDocToString(doc, options) {
         break;
       case DOC_TYPE_LINE_SUFFIX_BOUNDARY:
         if (lineSuffix.length > 0) {
-          cmds.push([ind, mode, { type: "line", hard: true }]);
```
Comment (thorn0) [inline src/document/printer.js]: It should be hardlineWithoutBreakParent

Comment (fisker) [reply src/document/printer.js]: Thanks

File: src/document/printer.js [hunk 2]
```diff
@@ -292,283 +317,292 @@ function printDocToString(doc, options) {
...
 
-    if (typeof doc === "string") {
-      const formatted = newLine !== "\n" ? doc.replace(/\n/g, newLine) : doc;
-      out.push(formatted);
-      pos += getStringWidth(formatted);
-    } else if (isConcat(doc)) {
-      const parts = getDocParts(doc);
-      for (let i = parts.length - 1; i >= 0; i--) {
-        cmds.push([ind, mode, parts[i]]);
+    switch (type) {
```
Comment (thorn0) [inline src/document/printer.js:322]: Let's do `switch(getDocType(doc))` instead?

Comment (fisker) [reply src/document/printer.js]: I was going to use type in default branch, but I don't need that anymore.

**Review by thorn0 — APPROVED** [review]
I'm not a fan of constants, but that's subjective.

**Comment by thorn0** [comment]
BTW, we're hitting TypeScript's another limitation here. `getDocType`-like functions can't narrow union types. There is an issue about that: https://github.com/microsoft/TypeScript/issues/30698

**Pull request merged**